src/main/java/org/opensearch/knn/index/codec/derivedsource/NestedPerFieldDerivedVectorInjector.java

```diff
@@ -0,0 +1,276 @@
+/*
+ * Copyright OpenSearch Contributors
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+package org.opensearch.knn.index.codec.derivedsource;
+
+import lombok.AllArgsConstructor;
+import lombok.extern.log4j.Log4j2;
+import org.apache.lucene.index.DocValuesType;
+import org.apache.lucene.index.FieldInfo;
+import org.apache.lucene.index.PostingsEnum;
+import org.apache.lucene.index.SegmentReadState;
+import org.apache.lucene.index.Terms;
+import org.apache.lucene.index.TermsEnum;
+import org.apache.lucene.search.DocIdSetIterator;
+import org.apache.lucene.util.BytesRef;
+import org.opensearch.index.mapper.FieldNamesFieldMapper;
+import org.opensearch.knn.index.vectorvalues.KNNVectorValues;
+import org.opensearch.knn.index.vectorvalues.KNNVectorValuesFactory;
+
+import java.io.IOException;
+import java.util.ArrayList;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+
+import static org.apache.lucene.search.DocIdSetIterator.NO_MORE_DOCS;
+
+@Log4j2
+@AllArgsConstructor
+public class NestedPerFieldDerivedVectorInjector implements PerFieldDerivedVectorInjector {
+
+    private final FieldInfo childFieldInfo;
+    private final DerivedSourceReaders derivedSourceReaders;
+    private final SegmentReadState segmentReadState;
+
+    @Override
+    public void inject(int parentDocId, Map<String, Object> sourceAsMap) throws IOException {
+        // If the parent has the field, then it is just an object field.
+        int lowestDocIdForFieldWithParentAsOffset = getLowestDocIdForField(childFieldInfo.name, parentDocId);
+        if (lowestDocIdForFieldWithParentAsOffset == parentDocId) {
+            injectObject(parentDocId, sourceAsMap);
+            return;
+        }
+
+        // Setup the iterator. Return if no parent
+        String childFieldName = ParentChildHelper.getChildField(childFieldInfo.name);
+        String parentFieldName = ParentChildHelper.getParentField(childFieldInfo.name);
+        if (parentFieldName == null) {
+            return;
+        }
```

Comment:
what will be the case when parentFieldName will be null?

Reply:
We shouldnt get here, but adding just as a check